eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelTypes.scala

```diff
@@ -105,34 +107,83 @@ case class BITCOIN_PARENT_TX_CONFIRMED(childTx: Transaction) extends BitcoinEven
        "Y8888P"   "Y88888P"  888       888 888       888 d88P     888 888    Y888 8888888P"   "Y8888P"
  */
 
-sealed trait Upstream
-object Upstream {
+/**
+ * Origin of a payment, answering both questions:
+ * - what actor in the app sent that htlc? (Origin.replyTo)
+ * - what are the parent(s) of this payment in the htlc chain? (Upstream)
+ */
+sealed trait Origin[+U <: Origin.Upstream] { def upstream: U }
+object Origin {
+  /** We have full details and the origin actor is known. */
+  sealed trait Hot[+U <: Upstream] extends Origin[U]
+  /** We have restarted after the payment was sent, we have limited info and the origin actor doesn't exist anymore */
+  sealed trait Cold[+U <: Upstream] extends Origin[U]
+
   /** Our node is the origin of the payment. */
-  final case class Local(id: UUID) extends Upstream
+  sealed trait Local extends Origin[Upstream.Local]
+  case class LocalHot(upstream: Upstream.Local, replyTo: ActorRef) extends Local with Hot[Upstream.Local]
+  case class LocalCold(upstream: Upstream.Local) extends Local with Cold[Upstream.Local]
+
   /** Our node forwarded a single incoming HTLC to an outgoing channel. */
-  final case class Relayed(add: UpdateAddHtlc) extends Upstream
-  /** Our node forwarded an incoming HTLC set to a remote outgoing node (potentially producing multiple downstream HTLCs). */
-  final case class TrampolineRelayed(adds: Seq[UpdateAddHtlc]) extends Upstream {
-    val amountIn: MilliSatoshi = adds.map(_.amountMsat).sum
-    val expiryIn: CltvExpiry = adds.map(_.cltvExpiry).min
+  sealed trait ChannelRelayed extends Origin[Upstream.ChannelRelayed]
+  case class ChannelRelayedHot(upstream: Upstream.ChannelRelayedHot, replyTo: ActorRef) extends ChannelRelayed with Hot[Upstream.ChannelRelayed]
+  case class ChannelRelayedCold(upstream: Upstream.ChannelRelayedCold) extends ChannelRelayed with Cold[Upstream.ChannelRelayed]
+
+  /** Our node forwarded an incoming HTLC set to a remote outgoing node (potentially producing multiple downstream HTLCs).*/
+  sealed trait TrampolineRelayed extends Origin[Upstream.TrampolineRelayed]
+  case class TrampolineRelayedHot(upstream: Upstream.TrampolineRelayedHot, replyTo: ActorRef) extends TrampolineRelayed with Hot[Upstream.TrampolineRelayed]
+  case class TrampolineRelayedCold(upstream: Upstream.TrampolineRelayedCold) extends TrampolineRelayed with Cold[Upstream.TrampolineRelayed]
+
+  /** Parents of a payment in the htlc chain */
+  sealed trait Upstream
+  object Upstream {
+    /** there is a payment lifecycle involved for local and trampoline payments */
+    sealed trait SentByPaymentLifecycle extends Upstream
```

Comment:
In fact I don't think this trait is useful at all. The only place where we really use it in the `SendPaymentConfig` of the `PaymentInitiator`, but using `Hot` there would be enough.
The only pattern matching we do on it is in `MultiPartPaymentLifecycle.spawnChildPaymentFsm` where we just need to do something for `Local` origin, and leave other kinds of origin unchanged.

Reply:
Actually, the reason why I introduced this abstraction is to avoid an impossible match/case here:

```scala
def buildCommand(replyTo: ActorRef, upstream: Upstream.SentByPaymentLifecycle, paymentHash: ByteVector32, hops: Seq[ChannelHop], finalPayload: Onion.FinalPayload): (CMD_ADD_HTLC, Seq[(ByteVector32, PublicKey)]) = {
    val (firstAmount, firstExpiry, onion) = buildPacket(Sphinx.PaymentPacket)(paymentHash, hops, finalPayload)
    val origin = upstream match {
      case u: Upstream.Local => Origin.LocalHot(u, replyTo)
      case u: Upstream.TrampolineRelayedHot => Origin.TrampolineRelayedHot(u, replyTo)
    }
    CMD_ADD_HTLC(replyTo, firstAmount, paymentHash, firstExpiry, onion.packet, origin, commit = true) -> onion.sharedSecrets
  }
```

Reply:
But is it really harmful to add `case u: Upstream.ChannelRelayedHot => Origin.ChannelRelayedHot(u, replyTo)` here? I don't have a very strong opinion, but I think the `Origin` type hierarchy is a bit complex already...

Reply:
It depends how hard we want to rely on the typing system to prevent us from making mistakes. Having a `Upstream.ChannelRelayedHot` is a bug that the compiler could detect.

Reply:
That's true, let's keep it then.

Reply:
The issue disappears with 4a25b39.